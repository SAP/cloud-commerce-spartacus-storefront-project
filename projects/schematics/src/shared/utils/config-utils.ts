import { Node, SourceFile, ts as tsMorph } from 'ts-morph';
import { PROVIDE_CONFIG_FUNCTION } from '../constants';
import { isImportedFromSpartacusLibs } from './import-utils';
import { getModule } from './new-module-utils';

export function getSpartacusProviders(sourceFile: SourceFile): Node[] {
  const module = getModule(sourceFile);
  if (!module) {
    return [];
  }
  const literal = module.getFirstDescendantByKind(
    tsMorph.SyntaxKind.ObjectLiteralExpression
  );
  const providers: Node[] = [];
  if (literal) {
    const properties =
      literal.getChildrenOfKind(tsMorph.SyntaxKind.PropertyAssignment) ?? [];
    properties.forEach((property) => {
      if (
        property.getNameNode().getText() === 'providers' &&
        property.getInitializerIfKind(tsMorph.SyntaxKind.ArrayLiteralExpression)
      ) {
        const initializer = property.getInitializerIfKind(
          tsMorph.SyntaxKind.ArrayLiteralExpression
        );
        initializer?.getElements().forEach((element) => {
          if (Node.isCallExpression(element) || Node.isSpreadElement(element)) {
            const expression = element.getExpression();
            if (
              Node.isIdentifier(expression) &&
              isImportedFromSpartacusLibs(expression)
            ) {
              providers.push(element);
            }
          }
        });
      }
    });
  }
  return providers;
}

const EMPTY_SPACE_REG_EXP = /\s+/gm;
export function normalizeConfiguration(config: string | Node): string {
  let newConfig = typeof config === 'string' ? config : config.getText();

  newConfig = newConfig.trim();

  if (newConfig.startsWith(PROVIDE_CONFIG_FUNCTION)) {
    newConfig = newConfig.replace(`${PROVIDE_CONFIG_FUNCTION}(`, '');
    newConfig = newConfig.substring(0, newConfig.length - 1);
  }

  newConfig = newConfig.replace(EMPTY_SPACE_REG_EXP, '');

  return newConfig;
}
