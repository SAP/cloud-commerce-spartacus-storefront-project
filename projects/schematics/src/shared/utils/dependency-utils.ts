import { SchematicsException } from '@angular-devkit/schematics';
import collectedDependencies from '../../dependencies.json';
import { CORE_SPARTACUS_SCOPES, SPARTACUS_SCOPE } from '../libs-constants';
import {
  getKeyByMappingValue,
  libraryFeatureMapping,
} from '../updateable-constants';
import {
  crossFeatureInstallationOrder,
  libraryInstallationOrder,
} from './graph-utils';
import { calculateSort } from './lib-utils';
import { getConfiguredDependencies } from './schematics-config-utils';

/**
 * Analyzes cross-feature Spartacus dependencies
 * for the given set of features.
 *
 * E.g. when installing Digital Payments feature,
 * the following features will also be configured:
 * Account, Profile, Cart, Order, Checkout
 *
 * Returns the ordered list, according to the graph.
 */
export function analyzeCrossFeatureDependencies(
  startingFeatures: string[]
): string[] {
  const result: string[] = [];

  for (const feature of startingFeatures) {
    collectCrossFeatureDeps(feature, result);
  }

  return result.sort((feature1, feature2) =>
    calculateSort(feature1, feature2, crossFeatureInstallationOrder)
  );
}

/**
 * Collects the cross-feature dependencies for the given feature.
 * It recursively collects the dependencies for each of the
 * found dependencies.
 */
function collectCrossFeatureDeps(feature: string, result: string[]): void {
  // already processed
  if (result.includes(feature)) {
    return;
  }

  const spartacusLib = getKeyByMappingValue(libraryFeatureMapping, feature);
  if (!spartacusLib) {
    throw new SchematicsException(
      `Spartacus library not found for '${spartacusLib}' in ${JSON.stringify(
        libraryFeatureMapping
      )}`
    );
  }

  result.push(feature);

  const featureDependencies = getConfiguredDependencies(feature);
  for (const featureDependency of featureDependencies) {
    collectCrossFeatureDeps(featureDependency, result);
  }
}

/**
 * Analyzes cross-library Spartacus dependencies
 * for the given set of features.
 *
 * For example, CDC depends on User and ASM features.
 *
 * Returns the ordered list, according to the graph.
 */
/**
 * TODO:#schematics - very similar to collectSpartacusLibraryDependencies() and collectSpartacusPeerDeps()
 * from projects/schematics/src/migrations/mechanism/dependency-management/dependency-management.ts
 */
export function analyzeCrossLibraryDependencies(
  startingFeatures: string[]
): string[] {
  const startingLibraries: string[] = [];
  for (const feature of startingFeatures) {
    const library = getKeyByMappingValue(libraryFeatureMapping, feature);
    if (!library) {
      throw new SchematicsException(
        `The given '${feature}' doesn't contain a Spartacus package mapping.
Please check its schematics configuration.`
      );
    }

    startingLibraries.push(library);
  }

  let spartacusPeerDeps: string[] = startingLibraries;
  for (const spartacusLib of startingLibraries) {
    collectCrossSpartacusPeerDeps(spartacusLib, spartacusPeerDeps);
  }

  // remove the duplicates
  spartacusPeerDeps = Array.from(new Set<string>(spartacusPeerDeps));
  // order the libraries
  spartacusPeerDeps = spartacusPeerDeps.sort((featureA, featureB) =>
    calculateSort(featureA, featureB, libraryInstallationOrder)
  );

  return spartacusPeerDeps;
}

/**
 * Recursively collects the cross Spartacus library dependencies for the given library.
 */
export function collectCrossSpartacusPeerDeps(
  name: string,
  collectedDeps: string[],
  processed: string[] = []
): void {
  if (processed.includes(name)) {
    return;
  }

  const peerDepsWithVersions =
    (collectedDependencies as Record<string, Record<string, string>>)[name] ??
    {};

  const peerDeps = Object.keys(peerDepsWithVersions)
    .filter((d) => d.startsWith(SPARTACUS_SCOPE))
    .filter((d) => !CORE_SPARTACUS_SCOPES.includes(d))
    .filter((d) => !collectedDeps.includes(d));

  collectedDeps.push(...peerDeps);
  processed.push(name);

  for (const peerDep of peerDeps) {
    collectCrossSpartacusPeerDeps(peerDep, collectedDeps);
  }
}
