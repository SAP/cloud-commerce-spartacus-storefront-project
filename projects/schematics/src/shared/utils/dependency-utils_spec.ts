/// <reference types="jest" />

import {
  CLI_CART_BASE_FEATURE,
  CLI_CDC_FEATURE,
  CLI_CHECKOUT_B2B_FEATURE,
  CLI_CHECKOUT_BASE_FEATURE,
  CLI_DIGITAL_PAYMENTS_FEATURE,
  CLI_ORDER_FEATURE,
  CLI_USER_ACCOUNT_FEATURE,
  CLI_USER_PROFILE_FEATURE,
  SPARTACUS_ASM,
  SPARTACUS_CART,
  SPARTACUS_CDC,
  SPARTACUS_CHECKOUT,
  SPARTACUS_CORE,
  SPARTACUS_DIGITAL_PAYMENTS,
  SPARTACUS_ORDER,
  SPARTACUS_ORGANIZATION,
  SPARTACUS_PRODUCT_CONFIGURATOR,
  SPARTACUS_STOREFRONTLIB,
  SPARTACUS_USER,
} from '../libs-constants';
import {
  analyzeCrossFeatureDependencies,
  analyzeCrossLibraryDependenciesByFeatures,
  analyzeCrossLibraryDependenciesByLibraries,
  collectCrossSpartacusPeerDeps,
} from './dependency-utils';

describe('dependency-util', () => {
  describe('analyzeCrossFeatureDependencies', () => {
    it('DP - should return the correct set of ordered sub-features', () => {
      const result = analyzeCrossFeatureDependencies([
        CLI_CDC_FEATURE,
        CLI_DIGITAL_PAYMENTS_FEATURE,
        CLI_CHECKOUT_B2B_FEATURE,
      ]);

      expect(result).toEqual([
        CLI_USER_ACCOUNT_FEATURE,
        CLI_USER_PROFILE_FEATURE,
        CLI_CART_BASE_FEATURE,
        CLI_ORDER_FEATURE,
        CLI_CHECKOUT_BASE_FEATURE,
        CLI_CHECKOUT_B2B_FEATURE,
        CLI_DIGITAL_PAYMENTS_FEATURE,
        CLI_CDC_FEATURE,
      ]);
    });
    it('User Profile - should return the correct set of ordered sub-features', () => {
      const result = analyzeCrossFeatureDependencies([
        CLI_USER_PROFILE_FEATURE,
      ]);

      expect(result).toEqual([
        CLI_USER_ACCOUNT_FEATURE,
        CLI_USER_PROFILE_FEATURE,
      ]);
    });
  });

  describe('analyzeCrossLibraryDependenciesByFeatures', () => {
    it('DP - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossLibraryDependenciesByFeatures([
        CLI_DIGITAL_PAYMENTS_FEATURE,
      ]);

      expect(result).toEqual([
        SPARTACUS_USER,
        SPARTACUS_CART,
        SPARTACUS_ORDER,
        SPARTACUS_CHECKOUT,
        SPARTACUS_DIGITAL_PAYMENTS,
      ]);
    });
    it('CDC - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossLibraryDependenciesByFeatures([
        CLI_CDC_FEATURE,
      ]);

      expect(result).toEqual([SPARTACUS_USER, SPARTACUS_ASM, SPARTACUS_CDC]);
    });
  });

  describe('analyzeCrossLibraryDependenciesByLibraries', () => {
    it('DP and CDC - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossLibraryDependenciesByLibraries([
        SPARTACUS_DIGITAL_PAYMENTS,
        SPARTACUS_CDC,
      ]);

      expect(result).toEqual([
        SPARTACUS_USER,
        SPARTACUS_CART,
        SPARTACUS_ORDER,
        SPARTACUS_CHECKOUT,
        SPARTACUS_ASM,
        SPARTACUS_DIGITAL_PAYMENTS,
        SPARTACUS_CDC,
      ]);
    });

    it('CDC - should return the correct set of ordered libraries', () => {
      const result = analyzeCrossLibraryDependenciesByLibraries([
        SPARTACUS_CDC,
      ]);

      expect(result).toEqual([SPARTACUS_USER, SPARTACUS_ASM, SPARTACUS_CDC]);
    });

    it('Core libs, Organization and Product configurator', () => {
      const result = analyzeCrossLibraryDependenciesByLibraries([
        SPARTACUS_CORE,
        SPARTACUS_STOREFRONTLIB,
        SPARTACUS_ORGANIZATION,
        SPARTACUS_PRODUCT_CONFIGURATOR,
      ]);

      expect(result).toEqual([
        SPARTACUS_USER,
        SPARTACUS_CART,
        SPARTACUS_ORDER,
        SPARTACUS_CHECKOUT,
        SPARTACUS_PRODUCT_CONFIGURATOR,
        SPARTACUS_ORGANIZATION,
        SPARTACUS_CORE,
        SPARTACUS_STOREFRONTLIB,
      ]);
    });
  });

  describe('collectCrossSpartacusPeerDeps', () => {
    it('should correctly collect all peer deps for the given library and its dependencies', () => {
      const result: string[] = [];
      collectCrossSpartacusPeerDeps(SPARTACUS_DIGITAL_PAYMENTS, result);

      expect(result).toContain(SPARTACUS_CART);
      expect(result).toContain(SPARTACUS_CHECKOUT);
      expect(result).toContain(SPARTACUS_ORDER);
      expect(result).toContain(SPARTACUS_USER);
    });
  });
});
