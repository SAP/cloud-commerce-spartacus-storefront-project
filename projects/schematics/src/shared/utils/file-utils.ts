import { experimental, strings } from '@angular-devkit/core';
import { SchematicsException, Tree } from '@angular-devkit/schematics';
import { getProjectTargetOptions } from '@angular/cdk/schematics';
import { parseTsconfigFile } from '@angular/core/schematics/utils/typescript/parse_tsconfig';
import { getSourceNodes } from '@schematics/angular/utility/ast-utils';
import {
  Change,
  InsertChange,
  ReplaceChange,
} from '@schematics/angular/utility/change';
import { dirname, relative } from 'path';
import * as ts from 'typescript';

export function getTsSourceFile(tree: Tree, path: string): ts.SourceFile {
  const buffer = tree.read(path);
  if (!buffer) {
    throw new SchematicsException(`Could not read file (${path}).`);
  }
  const content = buffer.toString();
  const source = ts.createSourceFile(
    path,
    content,
    ts.ScriptTarget.Latest,
    true
  );

  return source;
}

export function getAllTsSourceFiles(
  tsconfigPath: string,
  tree: Tree,
  basePath: string
): ts.SourceFile[] {
  const parsed = parseTsconfigFile(tsconfigPath, dirname(tsconfigPath));
  const host = createMigrationCompilerHost(tree, parsed.options, basePath);
  const program = ts.createProgram(parsed.fileNames, parsed.options, host);
  return program
    .getSourceFiles()
    .filter(
      f => !f.isDeclarationFile && !program.isSourceFileFromExternalLibrary(f)
    );
}

export function getIndexHtmlPath(
  project: experimental.workspace.WorkspaceProject
): string {
  const buildOptions = getProjectTargetOptions(project, 'build');

  if (!buildOptions.index) {
    throw new SchematicsException('"index.html" file not found.');
  }

  return buildOptions.index;
}

export function getPathResultsForFile(
  tree: Tree,
  file: string,
  directory?: string
): string[] {
  const results: string[] = [];
  const dir = directory || '/';

  tree.getDir(dir).visit(filePath => {
    if (filePath.endsWith(file)) {
      results.push(filePath);
    }
  });

  return results;
}

export enum InsertDirection {
  LEFT,
  RIGHT,
}

export function commitChanges(
  host: Tree,
  path: string,
  changes: Change[] | null,
  insertDirection: InsertDirection
): void {
  if (!changes) {
    return;
  }

  const recorder = host.beginUpdate(path);
  changes.forEach(change => {
    if (change instanceof InsertChange) {
      const pos = change.pos;
      const toAdd = change.toAdd;
      if (insertDirection === InsertDirection.LEFT) {
        recorder.insertLeft(pos, toAdd);
      } else {
        recorder.insertRight(pos, toAdd);
      }
    } else if (change instanceof ReplaceChange) {
      const pos = change['pos'];
      const oldText = change['oldText'];
      const newText = change['newText'];

      recorder.remove(pos, oldText.length);
      if (insertDirection === InsertDirection.LEFT) {
        recorder.insertLeft(pos, newText);
      } else {
        recorder.insertRight(pos, newText);
      }
    }
  });
  host.commitUpdate(recorder);
}

export function defineProperty(
  nodes: ts.Node[],
  path: string,
  toAdd: string
): InsertChange {
  const constructorNode = nodes.find(n => n.kind === ts.SyntaxKind.Constructor);

  if (!constructorNode) {
    throw new SchematicsException(`No constructor found in ${path}.`);
  }

  return new InsertChange(path, constructorNode.pos + 1, toAdd);
}

export function injectService(
  nodes: ts.Node[],
  path: string,
  serviceName: string,
  propertyName?: string
): InsertChange {
  const constructorNode = nodes.find(n => n.kind === ts.SyntaxKind.Constructor);

  if (!constructorNode) {
    throw new SchematicsException(`No constructor found in ${path}.`);
  }

  const parameterListNode = constructorNode
    .getChildren()
    .find(n => n.kind === ts.SyntaxKind.SyntaxList);

  if (!parameterListNode) {
    throw new SchematicsException(
      `No no parameter list found in ${path}'s constructor.`
    );
  }

  propertyName = propertyName
    ? strings.camelize(propertyName)
    : strings.camelize(serviceName);
  const toAdd = `private ${propertyName}: ${strings.classify(serviceName)}`;
  return new InsertChange(path, parameterListNode.pos, toAdd);
}

export function insertCommentAboveIdentifier(
  sourcePath: string,
  source: ts.SourceFile,
  identifierName: string,
  comment: string
): InsertChange[] {
  const callExpressionNodes = findNodesByTextAndKind(
    source,
    identifierName,
    ts.SyntaxKind.Identifier
  );
  const changes: InsertChange[] = [];
  callExpressionNodes.forEach(n =>
    changes.push(
      new InsertChange(
        sourcePath,
        getLineStartFromTSFile(source, n.getFullStart()),
        comment
      )
    )
  );
  return changes;
}

export function renameIdentifierNode(
  sourcePath: string,
  source: ts.SourceFile,
  oldName: string,
  newName: string
): ReplaceChange[] {
  const callExpressionNodes = findNodesByTextAndKind(
    source,
    oldName,
    ts.SyntaxKind.Identifier
  );
  const changes: ReplaceChange[] = [];
  callExpressionNodes.forEach(n =>
    changes.push(new ReplaceChange(sourcePath, n.getStart(), oldName, newName))
  );
  return changes;
}

function findNodesByTextAndKind(
  source: ts.SourceFile,
  text: string,
  syntaxKind: ts.SyntaxKind
): ts.Node[] {
  const nodes = getSourceNodes(source);
  return nodes
    .filter(n => n.kind === syntaxKind)
    .filter(n => n.getText() === text);
}

function getLineStartFromTSFile(
  source: ts.SourceFile,
  position: number
): number {
  const lac = source.getLineAndCharacterOfPosition(position);
  const lineStart = source.getPositionOfLineAndCharacter(lac.line, 0);

  return lineStart;
}

// as this is copied from https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/app-shell/index.ts#L211, no need to test Angular's code
export function getMetadataProperty(
  metadata: ts.Node,
  propertyName: string
): ts.PropertyAssignment {
  const properties = (metadata as ts.ObjectLiteralExpression).properties;
  const property = properties
    .filter(prop => prop.kind === ts.SyntaxKind.PropertyAssignment)
    .filter((prop: ts.PropertyAssignment) => {
      const name = prop.name;
      switch (name.kind) {
        case ts.SyntaxKind.Identifier:
          return (name as ts.Identifier).getText() === propertyName;
        case ts.SyntaxKind.StringLiteral:
          return (name as ts.StringLiteral).text === propertyName;
      }

      return false;
    })[0];

  return property as ts.PropertyAssignment;
}

// copied from https://github.com/angular/angular/blob/master/packages/core/schematics/utils/typescript/compiler_host.ts#L12, no need to test angular's code
export function createMigrationCompilerHost(
  tree: Tree,
  options: ts.CompilerOptions,
  basePath: string,
  fakeRead?: (fileName: string) => string | null
): ts.CompilerHost {
  const host = ts.createCompilerHost(options, true);

  // We need to overwrite the host "readFile" method, as we want the TypeScript
  // program to be based on the file contents in the virtual file tree. Otherwise
  // if we run multiple migrations we might have intersecting changes and
  // source files.
  host.readFile = fileName => {
    const treeRelativePath = relative(basePath, fileName);
    const fakeOutput = fakeRead ? fakeRead(treeRelativePath) : null;
    const buffer =
      fakeOutput === null ? tree.read(treeRelativePath) : fakeOutput;
    // Strip BOM as otherwise TSC methods (Ex: getWidth) will return an offset,
    // which breaks the CLI UpdateRecorder.
    // See: https://github.com/angular/angular/pull/30719
    return buffer ? buffer.toString().replace(/^\uFEFF/, '') : undefined;
  };

  return host;
}
