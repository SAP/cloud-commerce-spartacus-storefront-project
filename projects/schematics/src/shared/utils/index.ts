export {
  createNewConfig,
  getConfig,
  getExistingStorefrontConfigNode,
  mergeConfig,
} from './config-utils';
export {
  commitChanges,
  findMultiLevelNodesByTextAndKind,
  getTsSourceFile,
  InsertDirection,
  removeImport,
} from './file-utils';
export {
  addImport,
  addToMetadata,
  addToModuleImports,
  addToModuleImportsAndCommitChanges,
  addToModuleProviders,
  createImportChange,
} from './module-file-utils';
export {
  getSpartacusSchematicsVersion,
  readPackageJson,
} from './package-utils';
export {
  getDefaultProjectNameFromWorkspace,
  getProjectTargets,
  getSourceRoot,
  getWorkspace,
} from './workspace-utils';
