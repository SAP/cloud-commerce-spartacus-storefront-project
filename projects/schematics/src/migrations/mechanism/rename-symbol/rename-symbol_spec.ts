/// <reference types="jest" />

import { getSystemPath } from '@angular-devkit/core';
import { TempScopedNodeJsSyncHost } from '@angular-devkit/core/node/testing';
import { HostTree, Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as shx from 'shelljs';
import { runMigration, writeFile } from '../../../shared/utils/test-utils';

const MIGRATION_SCRIPT_NAME = 'migration-v4-rename-symbol-02';

const fileWithSimpleImport = `import { OtherComponent1, Test1Component } from "@spartacus/storefront";
import { Test2Component } from "@spartacus/core";

const array = [OtherComponent1, Test1Component, Test2Component];`;

// -----------------------------------------------------------------------

const fileWithSimpleImportWithAlias = `import { OtherComponent2 as Test, Test1Component } from "@spartacus/storefront";
import { Test2Component } from "@spartacus/core";

const array = [Test, Test1Component, Test2Component];`;

// -----------------------------------------------------------------------

const fileWithSimpleImportAndRename = `import { OtherComponent3, Test1Component } from "@spartacus/storefront";
import { Test2Component } from "@spartacus/core";

const array = [OtherComponent3, Test1Component, Test2Component];`;
// -----------------------------------------------------------------------

describe('renamed symbols', () => {
  let host: TempScopedNodeJsSyncHost;
  let appTree = Tree.empty() as UnitTestTree;
  let schematicRunner: SchematicTestRunner;
  let tmpDirPath: string;
  let previousWorkingDir: string;

  beforeEach(() => {
    schematicRunner = new SchematicTestRunner(
      'test',
      require.resolve('../../test/migrations-test.json')
    );
    host = new TempScopedNodeJsSyncHost();
    appTree = new UnitTestTree(new HostTree(host));

    writeFile(
      host,
      '/tsconfig.json',
      JSON.stringify({
        compilerOptions: {
          lib: ['es2015'],
        },
      })
    );

    writeFile(
      host,
      '/angular.json',
      JSON.stringify({
        projects: {
          'spartacus-test': {
            sourceRoot: 'src',
            architect: {
              build: { options: { tsConfig: './tsconfig.json' } },
            },
          },
        },
      })
    );

    previousWorkingDir = shx.pwd();
    tmpDirPath = getSystemPath(host.root);

    // Switch into the temporary directory path. This allows us to run
    // the schematic against our custom unit test tree.
    shx.cd(tmpDirPath);
  });

  afterEach(() => {
    shx.cd(previousWorkingDir);
    shx.rm('-r', tmpDirPath);
  });

  describe('Previous import', () => {
    it('should became new import', async () => {
      writeFile(host, '/src/index.ts', fileWithSimpleImport);

      await runMigration(appTree, schematicRunner, MIGRATION_SCRIPT_NAME);

      const content = appTree.readContent('/src/index.ts');
      expect(content).toMatchSnapshot();
    });

    it('should became new import and name with aliases', async () => {
      writeFile(host, '/src/index.ts', fileWithSimpleImportWithAlias);

      await runMigration(appTree, schematicRunner, MIGRATION_SCRIPT_NAME);

      const content = appTree.readContent('/src/index.ts');
      expect(content).toMatchSnapshot();
    });

    it('should became new import and new name', async () => {
      writeFile(host, '/src/index.ts', fileWithSimpleImportAndRename);

      await runMigration(appTree, schematicRunner, MIGRATION_SCRIPT_NAME);

      const content = appTree.readContent('/src/index.ts');
      expect(content).toMatchSnapshot();
    });
  });
});
