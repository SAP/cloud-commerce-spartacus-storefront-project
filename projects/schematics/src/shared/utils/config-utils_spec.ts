/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { InMemoryFileSystemHost, Project, SourceFile } from 'ts-morph';
import ts from 'typescript';
import {
  createNewConfig,
  getConfig,
  getConfigs,
  getExistingStorefrontConfigNode,
  getSpartacusProviders,
  mergeConfig,
} from './config-utils';
import { commitChanges, getTsSourceFile } from './file-utils';

const collectionPath = path.join(__dirname, '../../collection.json');
const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

// TODO:#10744 - cleanup after implementing the new config utils.
describe('Storefront config utils', () => {
  let appTree: UnitTestTree;
  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };
  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };
  const defaultOptions = {
    project: 'schematics-test',
  };
  const appModulePath = 'src/app/app.module.ts';

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync('add-spartacus', defaultOptions, appTree)
      .toPromise();
  });

  describe.skip('getExistingStorefrontConfigNode', () => {
    it('should get the Storefront config from app.module.ts file', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;

      expect(config).toBeTruthy();
      expect(config.getFullText()).toContain('B2cStorefrontModule.withConfig');
    });
  });

  describe.skip('getConfig', () => {
    it('should return the specified config from Storefront CallExpression AST node object', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(config, 'context');

      expect(currentContextConfig).toBeTruthy();
      expect(currentContextConfig?.getFullText()).toContain('currency:');
    });

    it('should return an undefined if the provided configName was not found', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const configByName = getConfig(config, 'test');

      expect(configByName).toBeFalsy();
      expect(configByName).toEqual(undefined);
    });
  });

  describe.skip('mergeConfig', () => {
    it('should merge the provided config array', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(
        config,
        'context'
      ) as ts.PropertyAssignment;
      const currencyChange = mergeConfig(
        appModulePath,
        currentContextConfig,
        'currency',
        ['EUR', 'JPY']
      );

      expect(appTree.readContent(appModulePath)).not.toContain('EUR');
      expect(appTree.readContent(appModulePath)).not.toContain('JPY');

      commitChanges(appTree, appModulePath, [currencyChange]);

      expect(appTree.readContent(appModulePath)).toContain('EUR');
      expect(appTree.readContent(appModulePath)).toContain('JPY');
    });

    it('should merge the provided regular config', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const backendConfig = getConfig(
        config,
        'backend'
      ) as ts.PropertyAssignment;

      const change = mergeConfig(appModulePath, backendConfig, 'occ', 'random');

      expect(appTree.readContent(appModulePath)).not.toContain('random');
      commitChanges(appTree, appModulePath, [change]);
      expect(appTree.readContent(appModulePath)).toContain('random');
    });

    it('should create a new config if there is nothing to be mergex', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(
        config,
        'context'
      ) as ts.PropertyAssignment;
      const baseSiteChange = mergeConfig(
        appModulePath,
        currentContextConfig,
        'urlParameters',
        ['baseSite', 'language', 'currency']
      );

      expect(appTree.readContent(appModulePath)).not.toContain(
        'urlParameters:'
      );

      commitChanges(appTree, appModulePath, [baseSiteChange]);

      expect(appTree.readContent(appModulePath)).toContain('urlParameters:');
    });
  });

  describe.skip('createNewConfig', () => {
    it('should nest the given new config in the given config object', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(
        config,
        'context'
      ) as ts.PropertyAssignment;
      const testConfigChange = createNewConfig(
        appModulePath,
        currentContextConfig,
        'testObjectConfig',
        ['value1', 'value2']
      );

      expect(appTree.readContent(appModulePath)).not.toContain(
        'testObjectConfig:'
      );

      commitChanges(appTree, appModulePath, [testConfigChange]);

      expect(appTree.readContent(appModulePath)).toContain('testObjectConfig:');
      expect(appTree.readContent(appModulePath)).toContain('value1');
      expect(appTree.readContent(appModulePath)).toContain('value2');
    });
  });

  describe('ts-morph config utils', () => {
    const configFileContent = `
import { NgModule } from '@angular/core';
import {
CartAddEntrySuccessEvent,
CartRemoveEntrySuccessEvent,
provideConfig,
} from '@spartacus/core';
import { NavigationEvent, defaultCmsContentProviders } from '@spartacus/storefront';
import { PersonalizationRootModule } from '@spartacus/tracking/personalization/root';
import { AepModule } from '@spartacus/tracking/tms/aep';
import { BaseTmsModule, TmsConfig } from '@spartacus/tracking/tms/core';
import { GtmModule } from '@spartacus/tracking/tms/gtm';
import { someFunction, someSpread } from '@some/package';

@NgModule({
imports: [
BaseTmsModule.forRoot(),
GtmModule,
AepModule,
PersonalizationRootModule,
],
providers: [
someFunction(),
...someSpread,
...defaultCmsContentProviders,
provideConfig(<TmsConfig>{
  tagManager: {
    gtm: {
      events: [NavigationEvent, CartAddEntrySuccessEvent],
    },
    aep: {
      events: [NavigationEvent, CartRemoveEntrySuccessEvent],
    },
  },
}),
provideConfig({
  featureModules: {
    personalization: {
      module: () =>
        import('@spartacus/tracking/personalization').then(
          (m) => m.PersonalizationModule
        ),
    },
  },
}),
],
})
export class TrackingFeatureModule {}
`;
    let sourceFile: SourceFile;

    beforeAll(() => {
      const project = new Project({
        fileSystem: new InMemoryFileSystemHost(),
      });
      sourceFile = project.createSourceFile('test.ts', configFileContent);
    });

    describe('getConfigs', () => {
      it('should return all configs from provideConfigs calls', () => {
        const configs = getConfigs(sourceFile);
        expect(configs.length).toEqual(2);
        expect(configs[0].getText()).toMatchSnapshot();
        expect(configs[1].getText()).toMatchSnapshot();
      });
    });

    describe('getSpartacusProviders', () => {
      it('should return all providers from spartacus in file', () => {
        const providers = getSpartacusProviders(sourceFile);
        expect(providers.length).toEqual(3);
        expect(providers[0].getText()).toMatchSnapshot();
        expect(providers[1].getText()).toMatchSnapshot();
        expect(providers[2].getText()).toMatchSnapshot();
      });
    });
  });
});
