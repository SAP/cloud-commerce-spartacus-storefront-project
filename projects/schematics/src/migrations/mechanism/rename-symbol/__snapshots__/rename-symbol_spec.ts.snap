// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renamed symbols Previous import should became new import 1`] = `
"import { Test2Component } from \\"@spartacus/core\\";
import { OtherComponent1 } from \\"@spartacus/storefinder/components\\";
import { Test1Component } from \\"@spartacus/storefront\\";

const array = [OtherComponent1, Test1Component, Test2Component];
"
`;

exports[`renamed symbols Previous import should became new import and name with aliases 1`] = `
"import { Test2Component } from \\"@spartacus/core\\";
import { OtherComponentTest2 as Test } from \\"@spartacus/storefinder/components\\";
import { Test1Component } from \\"@spartacus/storefront\\";

const array = [Test, Test1Component, Test2Component];
"
`;

exports[`renamed symbols Previous import should became new import and new name 1`] = `
"import { Test2Component } from \\"@spartacus/core\\";
import { OtherComponentTest3 } from \\"@spartacus/storefinder/components\\";
import { Test1Component } from \\"@spartacus/storefront\\";

const array = [OtherComponentTest3, Test1Component, Test2Component];
"
`;
